agent/ui_endpoint_test.go

```diff
@@ -415,3 +417,276 @@ func TestUiServices(t *testing.T) {
+	dump[0].Checks[0].RaftIndex = structs.RaftIndex{}
+	dump[0].GatewayService.RaftIndex = structs.RaftIndex{}
+	dump[1].GatewayService.RaftIndex = structs.RaftIndex{}
+
+	expect := structs.ServiceDump{
+		{
+			Node: &structs.Node{
+				Node:       "bar",
+				Address:    "127.0.0.2",
+				Datacenter: "dc1",
+			},
+			Service: &structs.NodeService{
+				ID:      "db",
+				Service: "db",
+				Weights: &structs.Weights{
+					Passing: 1,
+					Warning: 1,
+				},
+			},
+			Checks: structs.HealthChecks{
+				{
+					Node:        "bar",
+					CheckID:     types.CheckID("db-warning"),
+					Name:        "db-warning",
+					Status:      "warning",
+					ServiceID:   "db",
+					ServiceName: "db",
+				},
+			},
+			GatewayService: &structs.GatewayService{
+				Gateway:     structs.ServiceID{ID: "terminating-gateway"},
+				Service:     structs.ServiceID{ID: "db"},
+				GatewayKind: "terminating-gateway",
+			},
+		},
+		{
+			// Only GatewayService should be returned when linked service isn't registered
+			GatewayService: &structs.GatewayService{
+				Gateway:     structs.ServiceID{ID: "terminating-gateway"},
+				Service:     structs.ServiceID{ID: "redis"},
+				GatewayKind: "terminating-gateway",
+				CAFile:      "/etc/certs/ca.pem",
+				CertFile:    "/etc/certs/cert.pem",
+				KeyFile:     "/etc/certs/key.pem",
+			},
+		},
+	}
+	assert.ElementsMatch(t, expect, dump)
+}
+
+func TestUIGatewayServiceNodes_Ingress(t *testing.T) {
+	t.Parallel()
+
+	a := NewTestAgent(t, "")
+	defer a.Shutdown()
+
+	// Register ingress gateway and a service that will be associated with it
+	{
+		arg := structs.RegisterRequest{
+			Datacenter: "dc1",
+			Node:       "foo",
+			Address:    "127.0.0.1",
+			Service: &structs.NodeService{
+				ID:      "ingress-gateway",
+				Service: "ingress-gateway",
+				Kind:    structs.ServiceKindIngressGateway,
+				Port:    8443,
+			},
+			Check: &structs.HealthCheck{
+				Name:      "ingress connect",
+				Status:    api.HealthPassing,
+				ServiceID: "ingress-gateway",
+			},
+		}
+		var regOutput struct{}
+		require.NoError(t, a.RPC("Catalog.Register", &arg, &regOutput))
+
+		arg = structs.RegisterRequest{
+			Datacenter: "dc1",
+			Node:       "bar",
+			Address:    "127.0.0.2",
+			Service: &structs.NodeService{
+				ID:      "db",
+				Service: "db",
+			},
+			Check: &structs.HealthCheck{
+				Name:      "db-warning",
+				Status:    api.HealthWarning,
+				ServiceID: "db",
+			},
+		}
+		require.NoError(t, a.RPC("Catalog.Register", &arg, &regOutput))
+
+		// Register ingress-gateway config entry, linking it to db and redis (dne)
+		args := &structs.IngressGatewayConfigEntry{
+			Name: "ingress-gateway",
+			Kind: structs.IngressGateway,
+			Listeners: []structs.IngressListener{
+				{
+					Port:     8888,
+					Protocol: "tcp",
+					Services: []structs.IngressService{
+						{
+							Name: "db",
+						},
+					},
+				},
+				{
+					Port:     8080,
+					Protocol: "http",
```

Comment:
This will fail on the current `master` branch unless we also define a `service-defaults` config entry setting the protocol to `http` for `web` service.

We do that to make sure the protocols in the service graph all align properly, but it does become a bit of a hassle in tests...